### PR TITLE
CSV serialize

### PR DESCRIPTION
**Pull request by vladbmihaescu**
Adding class for serializing records into CSV format.

### CODE REVIEW

File: invenio_records_rest/serializers/csv.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            )
+            records.append(self.process_dict(processed_hit, filter_keys))
+
+        return self.format_csv(records)
+
+    def format_csv(self, records):
+        """Format list of flattened dictionaries into CSV format."""
+        headers = sorted(set([key for record in records for key in record]))
+
+        rows = [["\"" + str(record[key]) + "\"" if key in record else "" for key in headers] for
```
Comment (ntarocco) [inline invenio_records_rest/serializers/csv.py]: I think we can try to optimize this, maybe we can quickly discuss this together. IMHO you could try to build your `set` of keys for the headers while iterating and building the rows.

File: invenio_records_rest/config.py [hunk 2]
```diff
@@ -366,3 +366,6 @@ def deleted_pid_error_handler(error):
 
 RECORDS_REST_DEFAULT_RESULTS_SIZE = 10
 """Default search results size."""
+
+
+FILTER_KEYS = {}
```
Comment (ntarocco) [inline invenio_records_rest/config.py]: We need to find a better way for this config.

File: tests/test_serializer_csv.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    titles = fields.Raw(attribute='metadata.titles')
+    booklinks = fields.Raw(attribute='metadata.booklinks')
+    circulation = fields.Raw(attribute='metadata.circulation')
+    books = fields.Raw(attribute='metadata.books')
+    document_types = fields.Raw(attribute='metadata.document_types')
+    languages = fields.Raw(attribute='metadata.languages')
+    keywords = fields.Raw(attribute='metadata.keywords')
+    series_objs = fields.Raw(attribute='metadata.series_objs')
+
+    record_1 = {'titles': {"pid": 51, "other": 'A'},
```
Comment (ntarocco) [inline tests/test_serializer_csv.py]: can you please add one or multiple values to test corner cases, for example:
* something like `'Jane " Doe'`
* something like `"Jane ' Doe"`
* something like `'Jane "aka" Doe'`
* something like `'François Doe'`
* any other you can think of

File: invenio_records_rest/serializers/csv.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        return csv_string
+
+    def process_dict(self, dictionary, filter_keys):
+        """Returns a flattened dictionary as a string."""
+        flattened = {}
+
+        self.flatten(dictionary, flattened, filter_keys)
+        return flattened
+
+    def flatten(self, elem, flattened, filter_keys, parent_key='',
```
Comment (ntarocco) [inline invenio_records_rest/serializers/csv.py]: I am not an extreme fan of not pure functions, in particular with recursion, maybe we can see together if we can find a better way!

**Comment by ntarocco** [comment]
Good job! 👍 Please have a look to Travis too that is failing!

File: invenio_records_rest/serializers/csv.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                    del flattened[key]
+
+        return flattened
+
+    def flatten(self, elem, flattened, parent_key=''):
+        """Flattens nested dictionaries."""
+        if type(elem) is dict:
+            for key in elem:
+                self.flatten(elem[key], flattened, parent_key + key + '_')
+        elif type(elem) is list:
```
Comment (egabancho) [inline invenio_records_rest/serializers/csv.py]: I've always found more pythonic something like `isinstance(elem, MutableSequence)` but I guess it's just personal preference ☺️

File: invenio_records_rest/serializers/csv.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            index = 0
+            for item in elem:
```
Comment (egabancho) [inline invenio_records_rest/serializers/csv.py]: `for index, item in enumerate(elem)` ?

**Comment by ntarocco** [comment]
@vladbmihaescu @egabancho @slint optimized the code with in memory buffer for the CSV output and return a generator. It should be more efficient with large amount of results. Added tests.
I have tested with a REST client and it looks working fine.

File: invenio_records_rest/serializers/csv.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    def _format_csv(self, records):
+        """Return the list of records as a CSV string."""
+        # build a unique list of all records keys as CSV headers
+        headers = set()
+        for rec in records:
+            headers.update(rec.keys())
+
+        # write the CSV output in memory
+        output = StringIO()
```
Comment (slint) [inline invenio_records_rest/serializers/csv.py]: TBH, I prefer the second solution from [this SO thread](https://stackoverflow.com/a/32609533) (the one with the `Line` class), it looks a bit higher-level (i.e. no `.seek` and `.truncate` calls)

Comment (ntarocco) [reply invenio_records_rest/serializers/csv.py]: You are right, this is much better! Will change that! 👍

File: invenio_records_rest/serializers/csv.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    """CSV serializer for records.
+
+    Note: This serializer is not suitable for serializing large number of
+    records.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialize CSVSerializer.
+
+        :param csv_excluded_fields: list of complete paths of the fields that
```
Comment (lnielsen) [inline invenio_records_rest/serializers/csv.py]: I must admit I find it more intuitive to specify what to include instead of what to exclude, but that's my personal opinion. I.e. either you include all if the user didn't specify anything, or you include just what the user specfied.

**Comment by lnielsen** [comment]
Please also add tests with unicode characters (this is usually how I break the Python csv module)

**Comment by lnielsen** [comment]
Btw, please also ensure you add it to the documentation:
https://github.com/inveniosoftware/invenio-records-rest/blob/master/docs/api.rst

**Comment by ntarocco** [comment]
> Please also add tests with unicode characters (this is usually how I break the Python csv module) and check the import into e.g. Excel or other spreadsheet to ensure it loads nicely by default.

actually my fault about this, I asked him to do it previously, he did it and then I pushed again removing that by mistake :) @vladbmihaescu sorry!
